### PR TITLE
[WIP] Non intrusive cmis integration

### PR DESCRIPTION
**Pull request by joeribekker**
**Changes**

This work is done for Utrecht with a different approach than the current CMIS-integration which requires heavy refactoring on every API update.

NOTE: Replaces #311

### CODE REVIEW

File: src/openzaak/components/documenten/admin.py [hunk 1]
```diff
@@ -40,6 +41,24 @@ class GebruiksrechtenAdmin(AuditTrailAdminMixin, UUIDAdminMixin, admin.ModelAdmi
...
+            return False
+        else:
+            return super().has_add_permission(request)
+
+    def has_change_permission(self, request, obj=None):
+        if settings.CMIS_ENABLED:
+            return False
+        else:
+            return super().has_change_permission(request, obj)
+
```
Comment (joeribekker) [inline src/openzaak/components/documenten/admin.py:61]: Could you move this to a mixin class? This prevents the same code in `EnkelvoudigInformatieObjectAdmin` and `ObjectInformatieObjectAdmin`

File: src/openzaak/components/documenten/api/serializers.py [hunk 2]
```diff
@@ -141,7 +141,9 @@ class EnkelvoudigInformatieObjectHyperlinkedRelatedField(LengthHyperlinkedRelate
     """
 
     def get_url(self, obj, view_name, request, format):
-        obj_latest_version = obj.latest_version
+        obj_latest_version = obj.get_latest_version(self.parent)
+        if obj_latest_version is None:
+            return None
```
Comment (joeribekker) [inline src/openzaak/components/documenten/api/serializers.py:146]: Hmm, when is this the case? I didn't see a test for this or did I miss it?

File: src/openzaak/components/documenten/api/validators.py [hunk 3]
```diff
@@ -39,9 +41,13 @@ class InformatieObjectUniqueValidator:
 
     def __call__(self, context: OrderedDict):
         informatieobject = context["informatieobject"]
-        oios = informatieobject.objectinformatieobject_set.filter(
-            **{context["object_type"]: context["object"]}
-        )
+        if settings.CMIS_ENABLED:
+            # TODO understand why it doesn't find the document
```
Comment (joeribekker) [inline src/openzaak/components/documenten/api/validators.py:45]: Was this solved?

File: src/openzaak/utils/tests.py [hunk 4]
```diff
@@ -141,3 +146,29 @@ def mock_client(responses: dict):
         delattr(sys.modules["zds_client.tests.mocks"], name)
     finally:
         pass
+
+
+# TODO Remove
```
Comment (joeribekker) [inline src/openzaak/utils/tests.py:151]: TODO :)

File: src/openzaak/components/zaken/tests/test_zaakinformatieobjecten.py [hunk 5]
```diff
@@ -394,7 +396,359 @@ def test_delete(self):
         self.assertTrue(Zaak.objects.exists())
 
 
+@override_settings(CMIS_ENABLED=True)
```
Comment (joeribekker) [inline src/openzaak/components/zaken/tests/test_zaakinformatieobjecten.py:399]: Can you move the classes with `CMIS_ENABLED=True` tests to a file with the postfix `_cmis`, so in this case `test_zaakinformatieobjecten_cmis.py`?

File: src/openzaak/components/zaken/api/serializers/zaken.py [hunk 6]
```diff
@@ -564,7 +565,10 @@ def create(self, validated_data):
         # If it fails in any other way, we need to handle that by rolling back
         # the ZIO creation.
         try:
-            response = create_remote_oio(io_url, zaak_url)
+            if settings.CMIS_ENABLED:
+                oio = ObjectInformatieObject.objects.create(informatieobject=io_url, zaak=zaak_url, object_type='zaak')
+            else:
+                response = create_remote_oio(io_url, zaak_url)
```
Comment (joeribekker) [inline src/openzaak/components/zaken/api/serializers/zaken.py:571]: Same as mentioned earlier, to move the logic to `create_remote_oio`.

File: src/openzaak/components/documenten/tests/test_validation.py [hunk 7]
```diff
@@ -6,19 +6,18 @@
...
-from openzaak.utils.tests import JWTAuthMixin
+from openzaak.utils.tests import APITestCaseCMIS, JWTAuthMixin
 
 from ..constants import OndertekeningSoorten, Statussen
 from .factories import EnkelvoudigInformatieObjectFactory
 
 
 @override_settings(ALLOWED_HOSTS=["testserver"])
-class EnkelvoudigInformatieObjectTests(JWTAuthMixin, APITestCase):
+class EnkelvoudigInformatieObjectTests(JWTAuthMixin, APITestCaseCMIS):
```
Comment (joeribekker) [inline src/openzaak/components/documenten/tests/test_validation.py:20]: If I understand correctly, CMIS is enabled by default (at the moment), so all these tests switched from testing Django models to testing CMIS. Please copy this file with the postfix `_cmis` and set appropriate CMIS_ENABLED overrides on classes in this file, and on the copied file.

File: src/openzaak/components/besluiten/tests/admin/test_besluitinformatieobject_audittrail.py [hunk 8]
```diff
@@ -15,6 +15,7 @@
 from ..utils import get_operation_url
 
 
+@override_settings(CMIS_ENABLED=False)
```
Comment (joeribekker) [inline src/openzaak/components/besluiten/tests/admin/test_besluitinformatieobject_audittrail.py:18]: Please remove `CMIS_ENABLED=False` everywhere (and override_settings import if not used)

File: src/openzaak/components/besluiten/tests/models/test_bio_block_change.py [hunk 9]
```diff
@@ -44,3 +48,45 @@ def test_bulk_create(self):
         self.assertRaises(
             QueryBlocked, BesluitInformatieObject.objects.bulk_create, [bio]
         )
+
+
+@override_settings(CMIS_ENABLED=True)
```
Comment (joeribekker) [inline src/openzaak/components/besluiten/tests/models/test_bio_block_change.py:53]: Please move this class to a new file `test_bio_block_change_cmis.py`

File: src/openzaak/components/documenten/admin.py (hunk 1 above)
Comment (SilviaAmAm) [reply src/openzaak/components/documenten/admin.py:61]: Done!

File: src/openzaak/components/documenten/api/serializers.py (hunk 2 above)
Comment (SilviaAmAm) [reply src/openzaak/components/documenten/api/serializers.py:146]: This is needed when the parent of `EnkelvoudigInformatieObjectHyperlinkedRelatedField` is a serializer and the parent.instance is a queryset, so it doesn't have the `get_informatieobject` method. 
For example, test `openzaak.components.documenten.tests.test_auth_cmis.ExternalInformatieobjecttypeScopeTests.test_oio_list` fails if that snippet is removed.

File: src/openzaak/loaders.py [hunk 10]
```diff
@@ -1,4 +1,5 @@
 import json
+import traceback
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/loaders.py:2]: Done!

File: src/openzaak/conf/includes/base.py [hunk 11]
```diff
@@ -530,9 +531,11 @@
...
 JWT_EXPIRY = config("JWT_EXPIRY", default=3600)
 
-
 NLX_DIRECTORY_URLS = {
     NLXDirectories.demo: "https://directory.demo.nlx.io/",
     NLXDirectories.preprod: "https://directory.preprod.nlx.io/",
     NLXDirectories.prod: "https://directory.prod.nlx.io/",
 }
+
+CMIS_ENABLED = True
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/conf/includes/base.py:540]: :+1:

File: src/openzaak/components/besluiten/api/serializers.py [hunk 12]
```diff
@@ -217,6 +221,19 @@ def create(self, validated_data):
                 code="pending-relations",
             )
         else:
-            bio._objectinformatieobject_url = response["url"]
+            if settings.CMIS_ENABLED:
+                bio._objectinformatieobject_url = oio.get_url()
+            else:
+                bio._objectinformatieobject_url = response["url"]
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/components/besluiten/api/serializers.py:227]: Indeed

File: src/openzaak/components/besluiten/api/serializers.py [hunk 13]
```diff
@@ -205,7 +206,10 @@ def create(self, validated_data):
         # If it fails in any other way, we need to handle that by rolling back
         # the BIO creation.
         try:
-            response = create_remote_oio(io_url, besluit_url, "besluit")
+            if settings.CMIS_ENABLED:
+                oio = ObjectInformatieObject.objects.create(informatieobject=io_url, besluit=besluit_url, object_type='besluit')
+            else:
+                response = create_remote_oio(io_url, besluit_url, "besluit")
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/components/besluiten/api/serializers.py:212]: Made the changes suggested and used a "fake" response

File: src/openzaak/components/documenten/api/validators.py (hunk 3 above)
Comment (SilviaAmAm) [reply src/openzaak/components/documenten/api/validators.py:45]: yes

File: src/openzaak/utils/tests.py (hunk 4 above)
Comment (SilviaAmAm) [reply src/openzaak/utils/tests.py:151]: Removed!

File: src/openzaak/components/zaken/api/serializers/zaken.py (hunk 6 above)
Comment (SilviaAmAm) [reply src/openzaak/components/zaken/api/serializers/zaken.py:571]: Done

File: src/openzaak/components/besluiten/api/serializers.py [hunk 14]
```diff
@@ -14,7 +14,7 @@
 from vng_api_common.validators import IsImmutableValidator, validate_rsin
 
 from openzaak.components.documenten.api.fields import EnkelvoudigInformatieObjectField
-from openzaak.components.documenten.api.utils import create_remote_oio
+from openzaak.components.documenten.api.utils import create_remote_oio, get_absolute_url
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/components/besluiten/api/serializers.py:17]: While cleaning up I realised that the functions in `openzaak.components.zaken.api.utils` are never used (`src/openzaak/components/zaken/api/serializers/zaken.py` uses `from openzaak.components.documenten.api.utils import create_remote_oio`) so for now I simply removed those. I am not too sure though which general `api.utils` file you mean?

File: requirements/base.txt [hunk 15]
```diff
@@ -31,20 +32,24 @@ django-redis==4.10.0
...
 drf-writable-nested==0.4.3
 drf-yasg==1.16.0
 gemma-zds-client==0.13.0  # via vng-api-common, zgw-consumers
+httplib2==0.17.0          # via cmislib-maykin
 humanize==0.5.1
 idna==2.6                 # via requests
 inflection==0.3.1         # via drf-yasg
 iso-639==0.4.5            # via vng-api-common
+iso-8601==0.3.0           # via drc-cmis
+iso8601==0.1.12           # via cmislib-maykin
```
[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply requirements/base.txt:52]: Updated the `non-intrusive_cmis-integration` branch of `gemma-drc-cmis` to have `iso8601` as the requirement and removed `iso-8601` from the `requirements/base.txt` in `open-zaak`

File: src/openzaak/components/besluiten/tests/models/test_bio_block_change.py (hunk 9 above)
Comment (SilviaAmAm) [reply src/openzaak/components/besluiten/tests/models/test_bio_block_change.py]: Addressed all the general test-related comments mentioned at the beginning.

File: src/openzaak/components/besluiten/api/serializers.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (joeribekker) [reply src/openzaak/components/besluiten/api/serializers.py]: well, a newly created general one ;-)

[the remark this answers is not in the document]
Comment (SilviaAmAm) [reply src/openzaak/components/besluiten/api/serializers.py]: Created a new `src/openzaak/utils/api.py` file with the utility functions and removed the old `src/openzaak/components/documenten/api/utils.py`.

**Comment by sergei-maertens** [comment]
Using #626 with a rebased branch for this feature. Currently reviewing/cleaning up.

**Pull request closed without merging**